### PR TITLE
Simplifies Ajax on both the JS and PHP sides.

### LINKED ISSUE #102: Simplify Ajax
The Ajax processing, on both the JS and PHP sides, is overly complicated and contains a lot of repetitious code.

This needs to be cleaned up before we can merge into core.

PR coming shortly.

### PR DESCRIPTION
**Pull request by pbiron**
Fixes #102, #94

This simplifies Ajax:

1. on the PHP side, combines both `enable/disable` handlers into a single `toggle` handler
2. on the JS side, reduces it down to a single click handler
    * also removes all `HTML injection`.  that is, the markup for all "states" is output from PHP when the screen is originally rendered and the JS basically just shows/hides the relevant parts depending on whether it is `enabling` or `disabling`

It also addresses the "unexpected/quite annoying jump/shift" when the Ajax is processing, as reported in #94.  It does that with the proposed solution in https://github.com/WordPress/wp-autoupdates/issues/94#issuecomment-623570523

Also fixes several other WPCS-related issues and an a11y problem where `@aria-label` wasn't correct in one place.

If this gets accepted, then we can close #99.

### CODE REVIEW

**Comment by pbiron** [comment]
While this is ready to be reviewed, it's not quite ready to be merged.  It still needs at least 2 things:

1. delay showing the spinning icon for a short time, so that if the Ajax completes really quickly would be less "jumpiness", as described in https://github.com/WordPress/wp-autoupdates/issues/94#issuecomment-622572079
2. remove the reliance on `wpAjax.unserialize()`

I've got some other things to attend to, I'll get to those as soon as I can.

**Comment by pbiron** [comment]
@audrasjb Can you please review the way I did the strings in `wp_autoupdates_toggle_auto_updates()`?  I not sure they are done in a way that is helpful to translators.

File: js/wp-autoupdates.js [hunk 1]
```diff
@@ -1,286 +1,88 @@
...
+
+		$.post( ajaxurl, data )
+			.done( function( response ) {
+				if ( response.success ) {
+					// Update the counts in the enabled/disabled views if on on
+					// screen with a list table.
+					// TODO: If either count started out 0 the appropriate span won't
+					//       be there and hence won't be updated.
+					if ( 'themes' !== pagenow ) {
+						var enabled        = $( '.autoupdate_enabled span' ),
```
Comment (afercia) [inline js/wp-autoupdates.js]: JS coding standards: `var` statement should be at the beginning of the function https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#declaring-variables-with-var

File: js/wp-autoupdates.js [hunk 2]
```diff
@@ -1,286 +1,88 @@
...
+						anchor.removeClass( 'enable' ).addClass( 'disable' );
+						label.text( wp_autoupdates.disable );
+						parent.find( '.auto-update-time').removeClass( 'hidden' );
+					} else {
+						anchor.removeClass( 'disable' ).addClass( 'enable' );
+						label.text( wp_autoupdates.enable );
+						parent.find( '.auto-update-time').addClass( 'hidden' );
+					}
+
+					wp.a11y.speak( 'enable' === action ? wp_autoupdates.enabled : wp_autoupdates.disabled, 'polite' );
```
Comment (afercia) [inline js/wp-autoupdates.js]: `polite` is redundant, as it's the default value.

File: js/wp-autoupdates.js [hunk 3]
```diff
@@ -1,286 +1,88 @@
...
+				} else {
+					parent.find( '.auto-updates-error' ).removeClass( 'hidden' ).addClass( 'notice error' ).find( 'p' ).text( response.data.error );
+					wp.a11y.speak( response.data.error, 'polite' );
```
Comment (afercia) [inline js/wp-autoupdates.js]: `polite` is redundant, as it's the default value.

File: js/wp-autoupdates.js [hunk 4]
```diff
@@ -1,286 +1,88 @@
...
+				}
+			} )
+			.fail( function( response ) {
+				parent.find( '.auto-updates-error' ).removeClass( 'hidden' ).addClass( 'notice error' ).find( 'p' ).text( wp_autoupdates.auto_update_error );
+				wp.a11y.speak( wp_autoupdates.auto_update_error, 'polite' );
```
Comment (afercia) [inline js/wp-autoupdates.js]: `polite` is redundant, as it's the default value.

File: js/wp-autoupdates.js (hunk 1 above)
Comment (pbiron) [reply js/wp-autoupdates.js]: I'm planning on raising that during `Open Floor` of `devchat` tomorrow.  Yes, by the "letter of the law" having a `var` at the beginning of a block is "wrong"; but by the "spirit of the law" (i.e., prevent polluting the global scope) it is fine.  So, I think the JSCS should be modified.

Thoughts?

File: js/wp-autoupdates.js (hunk 2 above)
Comment (pbiron) [reply js/wp-autoupdates.js]: @audrasjb What are your thoughts here?

Personally, I think it is sometimes **useful** to be explicit with params, even when using the default value...and think this is one of those cases.

File: js/wp-autoupdates.js (hunk 1 above)
Comment (afercia) [reply js/wp-autoupdates.js]: I think it's not correct because JavaScript has something called "hoisting" 🙂 The `var` declarations will be "moved" at the top of their scope anyways during the compile phase. `var` is function-scoped. This is why the coding standards ask to have them at the top: to avoid unexpected bugs and make code look closer to the way it runs. When using ES2015 or newer, one could use `const` and `let` which are block-scoped instead.

Comment (pbiron) [reply js/wp-autoupdates.js]: Thanx, I wasn't aware of that!  Will certainly fix then.

File: js/wp-autoupdates.js [hunk 5]
```diff
@@ -1,286 +1,88 @@
...
+			} )
+			.always( function() {
+				anchor.find( '.dashicons-update' ).addClass( 'hidden' );
+			} );
+	} );
+} )( jQuery );
```
Comment (azaozz) [inline js/wp-autoupdates.js]: Just a nitpick but when merging to core perhaps better to pass `( window.jQuery, window.pagenow )` and run this on DOM ready.

File: js/wp-autoupdates.js (hunk 1 above)
Comment (azaozz) [reply js/wp-autoupdates.js]: Yeah the vars can be defined right above `if ( response.success ) {`. That is how it is executed in js. 
This doesn't apply to `const` and `let`, but they have different scope too.

**Review by azaozz — CHANGES_REQUESTED** [review]
Thinking this is ready to merge with couple of tiny changes/improvements. Can be refined more while or after the core patch if needed.

File: js/wp-autoupdates.js (hunk 5 above)
Comment (pbiron) [reply js/wp-autoupdates.js]: agree on pagenow, will fix that.

on jQuery, most of the core JS I've looked at just uses `jQuery`, e.g., `wp-admin/js/updates.js`)

**Comment by pbiron** [comment]
> Thinking this is ready to merge with couple of tiny changes/improvements. Can be refined more while or after the core patch if needed.

I'm just about done with replacing `wpAjax.unserialize()` with `@data` attributes.  Will push that shortly.

**Comment by pbiron** [comment]
@azaozz I think it can be merged now, and any other changes that are necessary can be done in a separate PR.

**Review by azaozz — APPROVED** [review]
